### ISSUE TITLE
Capture error messages from lexing and parsing in C/Ruby

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber-attic/bool#2]
The _Exceptions_ spec in [ruby/spec/bool_spec.rb](https://github.com/cucumber/bool/blob/master/ruby/spec/bool_spec.rb) doesn't make an assertion about the error messages, but it should. (It's commented out).

This is needed in order to make it easier for users to fix their typos.

I haven't quite figured out how to do this in C/Ruby, but I have a hunch that we need to use [strerror](http://www.gnu.org/software/libc/manual/html_node/Error-Messages.html) and store the error whenever `yyerror` is called.

@mattwynne and @sebrose what do you think about this as a warm-up challenge for gherkin3?

**Comment by aslakhellesoy** [comment]
@sebrose you mentioned you'd like to take a look at this ticket. Still up for it?

**Comment by sebrose** [comment]
Yes. I'm on it.

On 5 February 2013 14:51, Aslak Hellesøy notifications@github.com wrote:

> @sebrose https://github.com/sebrose you mentioned you'd like to take a
> look at this ticket. Still up for it?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/cucumber/bool/issues/2#issuecomment-13132475..

## 

ACCU - Professionalism in Programming - http://accu.org

http://www.claysnow.co.uk
http://twitter.com/#!/sebrose
http://uk.linkedin.com/in/sebrose

**Comment by sebrose** [comment]
2 things:

1) Pulled down latest from Github 7ef8e06 - now bool/c builds fine, but
bool/ruby fails because of warnings in lexer.c and ast.c -

Removing -Wall from extconf.rb fixes it, but can't see why it succeeds in
bool/c and not from bool/ruby.

2) When I have changes to commit am I to put them straight into
cucumber/bool, or shall I commit them on a fork and issue a pull?

Cheers
Seb

On 5 February 2013 14:53, Seb Rose seb@claysnow.co.uk wrote:

> Yes. I'm on it.
> 
> On 5 February 2013 14:51, Aslak Hellesøy notifications@github.com wrote:
> 
> > @sebrose https://github.com/sebrose you mentioned you'd like to take a
> > look at this ticket. Still up for it?
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/cucumber/bool/issues/2#issuecomment-13132475..
> 
> ## 
> 
> ACCU - Professionalism in Programming - http://accu.org
> 
> http://www.claysnow.co.uk
> http://twitter.com/#!/sebrose
> http://uk.linkedin.com/in/sebrose

## 

ACCU - Professionalism in Programming - http://accu.org

http://www.claysnow.co.uk
http://twitter.com/#!/sebrose
http://uk.linkedin.com/in/sebrose

**Comment by aslakhellesoy** [comment]
Are you on 7ef8e060483fcfbf1e089f6f1825b648b51401b3 ?
I'd like to keep -Wall if possible. The build passes on Travis, my ubuntu box and my OS X box.

While fixing #3 I made bison 2.7 a requirement. It's needed to avoid warnings. What flex and bison versions do you have, what is your C compiler (gcc or clang), what version is it and finally - what is the warning causing it to fail?

**Comment by mattwynne** [comment]
@seb re fixes, I wouldn't make your own fork. If you have a change that you think might need review / discussion by other people on the team, make a pull request from a branch within this repo, and then we can discuss the changes there.

The https://github.com/defunkt/hub project gives you loads of useful CLI tools for interacting with Github pull requests.

**Comment by mattwynne** [comment]
I have a failing Ruby build now too as of 7ef8e06. Here's the error I get:

```
➤ rake
make parser.h
make: `parser.h' is up to date.
make parser.c
make: `parser.c' is up to date.
cd tmp/x86_64-darwin11.3.0/bool_ext/1.9.3
make
compiling ../../../../ext/bool_ext/lexer.c
cc1: warnings being treated as errors
lexer.c: In function 'yy_fatal_error':
lexer.c:1568: warning: unused parameter 'yyscanner'
lexer.c: In function 'yyalloc':
lexer.c:1900: warning: unused parameter 'yyscanner'
lexer.c: In function 'yyrealloc':
lexer.c:1905: warning: unused parameter 'yyscanner'
lexer.c: In function 'yyfree':
lexer.c:1917: warning: unused parameter 'yyscanner'
make: *** [lexer.o] Error 1
rake aborted!
```

I have:
- flex 2.5.35 Apple(flex-31)
- bison 2.7
- gcc 4.2.1
- clang-425.0.24 / Apple LLVM version 4.2 (based on LLVM 3.2svn)

How do I tell which C compiler I'm using?

**Comment by sebrose** [comment]
On 5 February 2013 16:45, Matt Wynne notifications@github.com wrote:

> I have a failing Ruby build now too as of 7ef8e06https://github.com/cucumber/bool/commit/7ef8e06.
> Here's the error I get:
> 
> ➤ rake
> make parser.h
> make: `parser.h' is up to date.
> make parser.c
> make:`parser.c' is up to date.
> cd tmp/x86_64-darwin11.3.0/bool_ext/1.9.3
> make
> compiling ../../../../ext/bool_ext/lexer.c
> cc1: warnings being treated as errors
> lexer.c: In function 'yy_fatal_error':
> lexer.c:1568: warning: unused parameter 'yyscanner'
> lexer.c: In function 'yyalloc':
> lexer.c:1900: warning: unused parameter 'yyscanner'
> lexer.c: In function 'yyrealloc':
> lexer.c:1905: warning: unused parameter 'yyscanner'
> lexer.c: In function 'yyfree':
> lexer.c:1917: warning: unused parameter 'yyscanner'
> make: **\* [lexer.o] Error 1
> rake aborted!
> 
> I have:
> - flex 2.5.35 Apple(flex-31)
> - bison 2.7
> - gcc 4.2.1
> - clang-425.0.24 / Apple LLVM version 4.2 (based on LLVM 3.2svn)
> 
> How do I tell which C compiler I'm using?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/cucumber/bool/issues/2#issuecomment-13138787.

These are the same errors as me.

I'm using:
bison 2.7
flex 2.5.35
gcc 4.2.1 (though I appear to have 2 builds of this 5666dot3 and 5658 -
however neither of them complains when I compile from the terminal)
I don't believe I'm using clang

## 

ACCU - Professionalism in Programming - http://accu.org

http://www.claysnow.co.uk
http://twitter.com/#!/sebrose
http://uk.linkedin.com/in/sebrose

**Comment by aslakhellesoy** [comment]
@mattwynne can you try again after doing a `make clean` from the root?
You should can see what C compiler is used in the standard output and in your mkmf.log files under tmp.

**Comment by mattwynne** [comment]
OK now I get a new error:

```
➤ rake
cd tmp/x86_64-darwin11.3.0/bool_ext/1.9.3
make
compiling ../../../../ext/bool_ext/ast.c
cc1: warnings being treated as errors
../../../../ext/bool_ext/ast.c: In function 'yyerror':
../../../../ext/bool_ext/ast.c:5: warning: unused parameter 'scanner'
../../../../ext/bool_ext/ast.c:5: warning: unused parameter 'node'
../../../../ext/bool_ext/ast.c:5: warning: unused parameter 'msg'
make: *** [ast.o] Error 1
```

And it looks like I'm using gcc. Is that good? :)

**Comment by aslakhellesoy** [comment]
It's good that you're using gcc, because I'm using clang. That way we're covering more platforms. If you upgrade your XCode you'll get clang too, but it's possible to force gcc with `CONFIG['CC'] = 'gcc'` in `extconf.rb`.

I have a fix in mind based on http://stackoverflow.com/questions/3599160/unused-parameter-warnings-in-c-code, but we also need to inject some `UNUSED` into the flex generated `lexer.c` with `sed`. Flex is known not to generate warning free code.

We could of course use less warnings, but since most of us are C noobs I think it's a good thing to be strict.

**Comment by aslakhellesoy** [comment]
For now, just remove `-Werror` from `c/Makefile` and `extconf.rb` locally to get around the warning problems.

**Comment by sebrose** [comment]
I'm now getting the following in the Java build

[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR]
/Users/macbookair/Projects/bool/java/target/generated-sources/jflex/bool/JFlexLexer.java:[16,36]
interface expected here
[ERROR]
/Users/macbookair/Projects/bool/java/target/generated-sources/jflex/bool/JFlexLexer.java:[229,5]
method does not override or implement a method from a supertype
[ERROR]
/Users/macbookair/Projects/bool/java/target/generated-sources/jflex/bool/JFlexLexer.java:[234,5]
method does not override or implement a method from a supertype
[INFO] 3 errors

## 

ACCU - Professionalism in Programming - http://accu.org

http://www.claysnow.co.uk
http://twitter.com/#!/sebrose
http://uk.linkedin.com/in/sebrose

**Comment by sebrose** [comment]
Committed changes that capture lexing and parsing errors and report them in the exception raised.

Still need to get my head around Bison's %locations directive and all the changes that entails if we want to capture line/column number of error too.

**Comment by aslakhellesoy** [comment]
@sebrose I removed JFlexLexer a while ago. You need to do a clean build.

**Comment by sebrose** [comment]
Doh!

On 5 February 2013 21:27, Aslak Hellesøy notifications@github.com wrote:

> @sebrose https://github.com/sebrose I removed JFlexLexer a while ago.
> You need to do a clean build.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/cucumber/bool/issues/2#issuecomment-13152811.

## 

ACCU - Professionalism in Programming - http://accu.org

http://www.claysnow.co.uk
http://twitter.com/#!/sebrose
http://uk.linkedin.com/in/sebrose

**Comment by aslakhellesoy** [comment]
I see you fixed this @sebrose, good job. For the future, when you fix a ticket, close it with a commit message so we can trace what fixed it later. This can be done with a commit message containing "Closes #2,#3". Don't remember exactly where on github this is documented. It's worth looking up.

**Comment by sebrose** [comment]
Will do.
